src/app/test_executive/slot_end_test.ml

```diff
@@ -0,0 +1,203 @@
+    in
+    let slot_tx_end = Mina_numbers.Global_slot.of_int slot_tx_end in
+    let slot_chain_end = Mina_numbers.Global_slot.of_int slot_chain_end in
+    let%bind () =
+      section_hard "spawn transaction sending"
+        (let num_payments = num_slots * window_ms / tx_delay_ms in
+         let repeat_count = Unsigned.UInt32.of_int num_payments in
+         let repeat_delay_ms = Unsigned.UInt32.of_int tx_delay_ms in
+         let num_sender_keys = List.length sender_priv_keys in
+         let n_bp_senders = List.length bp_senders in
+         let keys_per_sender = num_sender_keys / n_bp_senders in
+         [%log info]
+           "will now send %d payments from as many accounts. %d nodes will \
+            send %d payments each from distinct keys"
+           num_payments n_bp_senders keys_per_sender ;
+         Malleable_error.List.fold ~init:sender_priv_keys bp_senders
+           ~f:(fun keys node ->
+             let keys0, rest = List.split_n keys keys_per_sender in
+             Integration_test_lib.Graphql_requests.must_send_test_payments
+               ~repeat_count ~repeat_delay_ms ~logger ~senders:keys0
+               ~receiver_pub_key ~amount ~fee
+               (Network.Node.get_ingress_uri node)
+             >>| const rest )
+         >>| const () )
+    in
+    let%bind () =
+      section
+        (Printf.sprintf "wait until slot %d" num_slots)
+        Async.(at end_t >>= const Malleable_error.ok_unit)
+    in
+    let ok_if_true s =
+      Malleable_error.ok_if_true ~error:(Error.of_string s) ~error_type:`Soft
+    in
+    let%bind blocks =
+      Integration_test_lib.Graphql_requests
+      .must_get_best_chain_for_slot_end_test ~max_length:(2 * num_slots) ~logger
+        (Network.Node.get_ingress_uri receiver)
+    in
+    let%bind () =
+      section "blocks produced before slot_tx_end"
+        ( ok_if_true "only empty blocks were produced before slot_tx_end"
+        @@ List.exists blocks ~f:(fun block ->
+               Mina_numbers.Global_slot.(
+                 of_uint32 block.slot_since_genesis < slot_tx_end)
+               && ( block.command_transaction_count <> 0
+                  || block.snark_work_count <> 0
+                  || block.coinbase <> 0 ) ) )
+    in
+    let%bind () =
+      section "blocks produced after slot_tx_end"
+        (Malleable_error.List.iter blocks ~f:(fun block ->
+             let msg =
+               Printf.sprintf
+                 "non-empty block after slot_tx_end. block slot since genesis: \
+                  %s, txn count: %d, snark work count: %d, coinbase: %d"
+                 (Mina_numbers.Global_slot.to_string block.slot_since_genesis)
+                 block.command_transaction_count block.snark_work_count
+                 block.coinbase
+             in
+             ok_if_true msg
+               ( Mina_numbers.Global_slot.(
+                   of_uint32 block.slot_since_genesis < slot_tx_end)
+               || block.command_transaction_count = 0
```

Comment:
NIT: I'd put conjunction group into braces for clarity (not necessary, but required me to remember operator precedence to be sure it works)

Reply:
I certainly agree. It just turns out the formatter doesn't let me keep them :)